### PR TITLE
Virtual heatpump

### PR DESCRIPTION
**Pull request by spyrostz**
## Reason for the proposed changes

Please describe what we want to achieve and why.

## Proposed changes

-

**INTEGRATION_TESTS_BRANCH**=master
**GSY_FRAMEWORK_BRANCH**=feature/GSYE-636

### CODE REVIEW

File: src/gsy_e/models/strategy/energy_parameters/heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/heat_pump.py:211]: I think that you can remove this from here, as the two remaining calls do not depend on each other.

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:18]:
```suggestion
WATER_SPECIFIC_HEAT_CAPACITY = 4182  # [J/kg°C]
```

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:36]: IIRC, this is given in m^3/h, so how about:
```suggestion
            dh_water_flow_m3_h_profile: Optional[Union[str, float, Dict]] = None,
```

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:92]: Thanks for this assert! Maybe something for an enhancement story:
We should somehow shield the user from running into this and validate the input profiles for the fact that the supply_temp is always grater than the return temperature.What do you think?

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:81]: In order to not get confused later on, how about renaming `max_energy_consumption` into `min_energy_consumption` ?

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:102]: I think this method deserves a docstring even though it is private.
How about something like:
```python
"""
return the energy, that is needed to fill the heat demand which is equivalent to the temperature drop from supply to return temperature
"""
```
(did I get it right?)

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:117]: NIT: could be placed directly after line 113.

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:102]: Also, how about renaming `storage_temp` into `target_storage_temp_C` ?
Maybe also the method name could include `target` ?

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:144]: Also here I would like a docstring for later. For example:
```python
"""
Return the temperature of the storage after the provided energy amount is converted into heat.
"""
```

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:90]: Ignore me, the tank volume is provided it `l`, so all good.

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:17]: NIT:
```suggestion
COP_CALIBRATION_COEFFICIENT = 0.85
```

File: src/gsy_e/models/strategy/energy_parameters/heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/heat_pump.py:32]: Fixing, thanks!

File: src/gsy_e/models/strategy/state/heat_pump_state.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/state/heat_pump_state.py:80]:
```suggestion
        """Update the energy consumption of the heatpump for a given time slot."""
```

File: src/gsy_e/models/strategy/energy_parameters/heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/heat_pump.py:211]: Remnant from debugging, removing

File: src/gsy_e/models/strategy/virtual_heatpump.py
Comment (hannesdiedrich) [inline src/gsy_e/models/strategy/virtual_heatpump.py:14]: Nice how inheritance can save you from writing a lot of code. :-)

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:18]: Again correcting the experts, WATER_SPECIFIC_TEMPERATURE was the term used on Mustafa's document but you are obviously correct.

File: src/gsy_e/models/strategy/energy_parameters/heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/heat_pump.py:71]: Whoever merges second can do it :D I do not mind adapting this too

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:36]: Nop, this is accumulated m^3 / 15 mins. I have already converted the water flow profiles in research_scripts repo, and the profiles that you see used in the setup file are m3.

**Review by hannesdiedrich — COMMENTED** [review]
Some rather minor comments. LGTM in general and after addressed

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:92]: Hmmmm, let me add it even now. It is probably easy to add such a check in the market_cycle method.

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:81]: Agreed and renaming

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:102]: I left a comment, this is actually the energy required to heat the water and bring the water tank to storage_temp degrees celcius

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:117]: Fixed

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:95]: I am going to follow this however I am not that convinced that it is better that way. The temp increase / decrease might not be just from trading in the long run if we include DH. Will change it though and also on the base classes.

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:144]: Done

File: src/gsy_e/models/strategy/state/heat_pump_state.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/state/heat_pump_state.py:80]: Fixed

File: src/gsy_e/models/strategy/heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/heat_pump.py:111]: I do not want to actually call this method from the `VirtualHeatpumpStrategy` class constructor, and just call `self._init_price_params`. Reason for that, is that in order to do it from the `VirtualHeatpumpStrategy` class, I have to do `super().super().__init__()` which I find it uglier than the pylint error.

File: src/gsy_e/models/strategy/energy_parameters/heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/energy_parameters/heat_pump.py]: Ah I see now, that you either need to pass or add a docstring instead.
Alright. Next time I will not complain ;-)

File: src/gsy_e/models/strategy/energy_parameters/heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/energy_parameters/heat_pump.py:210]: This is still there.

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:92]: I do not see this being implemented. If you agree, let me add a story.

File: src/gsy_e/models/strategy/heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/heat_pump.py:111]: Ah ok, I totally agree that `super().super().__init__()` looks ugly. Ignore me.

**Review by hannesdiedrich — DISMISSED** [review]
Two comments are not addressed. LGTM already though

File: src/gsy_e/models/strategy/energy_parameters/heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/heat_pump.py:210]: Missed this, removing

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:92]: It is at the very bottom of this file:
https://github.com/gridsingularity/gsy-e/pull/1695/commits/2d3e9941b45ba6eaaab4e350747391cdd857d126#diff-0ed9e7c23f93c5256ee1bb8c050f874ab97a9164d637c73d289bee266fdbe546R216

**Comment by hannesdiedrich** [comment]
I am sorry, I did not think about the merge conflicts when hitting the merge button.

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:92]: Ah, thanks. What I really meant was to validate the profiles when reading (or uploading in the gsy-web case) in order to not let the simulation start and then get surprised by a failing simulation later. What do you think?

Comment (spyrostz) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:92]: Hmmm, this needs to be done in the gsy-web then and I do not think that I have an open PR here. As a matter of fact, this strategy is only supposed to live in gsy-e. So I am not sure whether something needs to be done here

**Review by hannesdiedrich — APPROVED** [review]
LGTM

File: src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py
Comment (hannesdiedrich) [reply src/gsy_e/models/strategy/energy_parameters/virtual_heat_pump.py:92]: It could be beneficial on the gsy-e side, too, IMO Because this would save us some re-runs of the simulation if this condition is not valid for the provided profile files (worst case at the end of the simulation time).
Again, nothing urgent, but IMO very good to have.

**Pull request merged**
